### ISSUE TITLE
what augumentation did you use? why i fond some issues?

### ISSUE THREAD

**Issue by wuting3** [google/automl#449]
I fond that you made a mistake when resize_and_crop_boxes 
you use     
boxes = preprocessor.box_list_scale(
        boxlist, self._scaled_height, self._scaled_width).get()
but i think you should change it to 
    boxes = preprocessor.box_list_scale(
        boxlist, self._image_scale, self._image_scale).get()

in addition
when you clip box, you use
def clip_boxes(self, boxes):
  """Clip boxes to fit in an image."""
  boxes = tf.where(tf.less(boxes, 0), tf.zeros_like(boxes), boxes)
  boxes = tf.where(tf.greater(boxes, self._output_size[0] - 1),
                   (self._output_size[1] - 1) * tf.ones_like(boxes), boxes)
 
That only can be used for rectangle that width=height,you can change it to :
  def clip_boxes(self, boxes):
    """Clip boxes to fit in an image."""
    y_min, x_min, y_max, x_max = tf.split(value=boxes, num_or_size_splits=4, axis=1)
    y_min_clipped = tf.maximum(tf.minimum(tf.squeeze(y_min), self._output_size[0]), 0)
    y_max_clipped = tf.maximum(tf.minimum(tf.squeeze(y_max), self._output_size[0]), 0)
    x_min_clipped = tf.maximum(tf.minimum(tf.squeeze(x_min), self._output_size[1]), 0)
    x_max_clipped = tf.maximum(tf.minimum(tf.squeeze(x_max), self._output_size[1]), 0)
    boxes = tf.stack([y_min_clipped,x_min_clipped,y_max_clipped,x_max_clipped],1)
    return boxes

more over
you did only "Filter out ground truth boxes that are all zeros.",but not "Filter out ground truth boxes that are all [self._scaled_height, self._scaled_width,self._scaled_height, self._scaled_width].",

so what argumentation did you use when you train?

**Comment by wuting3** [comment]
In addition
you did only "Filter out ground truth boxes that are all zeros.",but not "Filter out ground truth boxes that are all [self._scaled_height, self._scaled_width,self._scaled_height, self._scaled_width].",

**Comment by wuting3** [comment]
def resize_and_crop_boxes(self):
    """Resize boxes and crop it to the self._output dimension."""
    boxlist = preprocessor.box_list.BoxList(self._boxes)
    boxes = preprocessor.box_list_scale(
        boxlist, self._image_scale, self._image_scale).get()
    # Adjust box coordinates based on the offset.
    box_offset = tf.stack([self._crop_offset_y, self._crop_offset_x,
                           self._crop_offset_y, self._crop_offset_x,])
    boxes -= tf.cast(tf.reshape(box_offset, [1, 4]), tf.float32)

    # Filter out ground truth boxes that are out of window [0,0,self._output_size[0],self._output_size[1]].
    y_min, x_min, y_max, x_max = tf.split(
        value=boxes, num_or_size_splits=4, axis=1)
    coordinate_violations = tf.concat([
      tf.greater_equal(y_min, self._output_size[0]),
      tf.greater_equal(x_min, self._output_size[1]),
      tf.less_equal(y_max, 0),
      tf.less_equal(x_max, 0)], 1)
    valid_indices = tf.reshape(
        tf.where(tf.logical_not(tf.reduce_any(coordinate_violations, 1))), [-1])
    boxes = self.clip_boxes(boxes)
    boxes = tf.gather(boxes, valid_indices)
    classes = tf.gather(self._classes, valid_indices)
    self._boxes=boxes
    return boxes, classes

**Comment by mingxingtan** [comment]
Hi @wuting3 , thanks for pointing out these issues.  Fixes on the way.

- The first issue might has some effect: evaluating now.  (it is obviously a bug...)
- The second issue is not a problem for COCO because I always use images with height=width, but it could impact others.  We will fix that.
- The third issue may have limited impact (I would guess x_min=x_max or y_min=y_max is rare), but we will try to fix it later.

All released models are trained using the same open source code as here, and you are right, there are still bugs :)

**Comment by mingxingtan** [comment]
Forgot to update this issue:

- Issue 1 is expected: I added the comment here: https://github.com/google/automl/blob/b4489cf0d934ba4d7c7639329909366985cf9a49/efficientdet/dataloader.py#L167
- Issue 2 and 3 are fixed now.
